tests/test_eigen_tensor.py

```diff
@@ -0,0 +1,230 @@
+import sys
+
+import pytest
+
+np = pytest.importorskip("numpy")
+eigen_tensor = pytest.importorskip("pybind11_tests.eigen_tensor")
+
+submodules = [eigen_tensor.c_style, eigen_tensor.f_style]
+
+tensor_ref = np.empty((3, 5, 2), dtype=np.int64)
+
+for i in range(tensor_ref.shape[0]):
+    for j in range(tensor_ref.shape[1]):
+        for k in range(tensor_ref.shape[2]):
+            tensor_ref[i, j, k] = i * (5 * 2) + j * 2 + k
+
+indices = (2, 3, 1)
+
+
+def assert_equal_tensor_ref(mat, writeable=True, modified=0):
+    assert mat.flags.writeable == writeable
+
+    copy = np.array(tensor_ref)
+    if modified != 0:
+        copy[indices] = modified
+
+    np.testing.assert_array_equal(mat, copy)
+
+
+@pytest.mark.parametrize("m", submodules)
+@pytest.mark.parametrize("member_name", ["member", "member_view"])
+def test_reference_internal(m, member_name):
+    if not hasattr(sys, "getrefcount"):
+        pytest.skip("No reference counting")
+    foo = m.CustomExample()
+    counts = sys.getrefcount(foo)
+    mem = getattr(foo, member_name)
+    assert_equal_tensor_ref(mem, writeable=False)
+    new_counts = sys.getrefcount(foo)
+    assert new_counts == counts + 1
+    assert_equal_tensor_ref(mem, writeable=False)
+    del mem
+    assert sys.getrefcount(foo) == counts
```

Comment:
gc_collect is necessary to support non-reference count based interpreters like pypy. But this test doesn't support pypy anyways. (Ideally it would be rewritten to support pypy but that seems like it would be quite tricky).

I'll add the line anyways just for consistency.